python/ray/tune/callback.py

```diff
@@ -43,11 +47,36 @@ def train(config):
 
     """
 
-    def setup(self):
+    # arguments here match Experiment.public_spec
+    def setup(self,
+              stop: Optional["Stopper"] = None,
+              resources_per_trial: Union[None, Mapping[str, Union[
+                  float, int, Mapping]], "PlacementGroupFactory"] = None,
+              num_samples: Optional[int] = None,
+              **info):
         """Called once at the very beginning of training.
 
         Any Callback setup should be added here (setting environment
         variables, etc.)
+
+        Arguments:
+            stop (dict | callable | :class:`Stopper`): Stopping criteria.
+                If ``time_budget_s`` was passed to ``tune.run``, a
+                ``TimeoutStopper`` will be passed here, either by itself
+                or as a part of a ``CombinedStopper``.
+            resources_per_trial (dict|PlacementGroupFactory): Machine resources
+                to allocate per trial, e.g. ``{"cpu": 64, "gpu": 8}``.
+                Note that GPUs will not be assigned unless you specify them
+                here. Defaults to 1 CPU and 0 GPUs in
+                ``Trainable.default_resource_request()``. This can also
+                be a PlacementGroupFactory object wrapping arguments to
+                create a per-trial placement group.
+            num_samples (int): Number of times to sample from the
+                hyperparameter space. Defaults to 1. If `grid_search` is
+                provided as an argument, the grid will be repeated
+                `num_samples` of times. If this is -1, (virtually) infinite
+                samples are generated until a stopping condition is met.
```

Comment:
Can you give some details why these arguments are required or what they will be used for? Especially the num samples can be quite misleading - for grid search the real number of samples will require calculation. The resources per trial can be fetched from the trial objects themselves, why do we need this in a setup function?

Reply:
`resources_per_trial` are superfluous, I agree. `num_samples` and `stop` are intended to provide necessary information for deadline-based callbacks (eg. a progress bar would require the maximum number of samples to display properly).

Reply:
Can we get rid of the resources per trial then?

Reply:
Another question: Should `num_samples` then be the total number of samples after grid searching, or is the `num_samples` parameter sufficient?

Reply:
It would be best if we could get the total number of samples, but that would need to be passed later, right?

Reply:
There is this line:

```
    progress_reporter.set_total_samples(search_alg.total_samples)
```

Can we just use `search_alg.total_samples` and rename the parameter to `total_num_samples` (to avoid confusion with `num_samples`)?